### PR TITLE
Implemented changes for chef extension commands in ARM mode 

### PR DESCRIPTION
**Pull request by siddheshwar-more**
Changes for chef extension related commands in ARM mode: 

```
1. vm extension set-chef [options] <resource-group> <vm-name> 
2. vm extension get-chef [options] <resource-group> <vm-name>
```

these commands will help user to perform chef extension related operation on given resource groups vm. 

Note- changes are in progress

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@siddheshwar-more__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, AZPRBOT;

**Comment by amarzavery** [comment]
@siddheshwar-more  - What is the status of this PR? Can you please fix the failing CI or should we close this PR?

**Comment by siddheshwar-more** [comment]
@amarzavery  

I've added tests for above changes but I'm facing issues while recording tests.

While running tests its giving me- 
```
{ text: '',
  errorText: 'error: \'set-chef\' is not an azure command. See \'azure help\'.\nerror: \'get-chef\' is not an azure command. See \'azure help\'.\n',
  exitStatus: 0 }

```

seems context issue with tests and source code!

Do you have any idea about this ?

**Comment by siddheshwar-more** [comment]
@amarzavery 

I've tried installation from source on fresh VM- 
https://github.com/Azure/azure-xplat-cli#download-source-code 

But no luck, even `bin/azure vm extension list --help` not lists `set-chef` and `get-chef` commands.

This is happening only for ARM mode. Is there any configuration required for ARM mode apart from `azure config mode arm` ?

Could you please help ?

**Comment by yugangw-msft** [comment]
please run "node bin/azure --gen" to refresh the metadata file, then your issues will be resolved

**Comment by siddheshwar-more** [comment]
@yugangw-msft  - After `"node bin/azure --gen"` execution `set-chef` and `get-chef` available for test context. Thank You!

Now facing one issue - test case fails even if test exit status is 0. i.e ` result.exitStatus.should.equal(0);`
 
```
     it('Set Chef extension for the created vm', function(done) {
       var cmd = util.format('vm extension set-chef %s %s --version %s --client-config %s --validation-pem %s --json', groupName, vmPrefix, '1210.12', clientConfig, validationPem).split(' ');
        testUtils.executeCommand(suite, retry, cmd, function(result) {
          result.exitStatus.should.equal(0);
          done();
        });
      });

```

**Comment by yugangw-msft** [comment]
@siddheshwar-more, I am pasting the test failure. Please record your test. Start with [this document](https://github.com/Azure/azure-xplat-cli/blob/dev/Documentation/Writing-Test.md)

Error: Cannot find module '/home/travis/build/Azure/azure-xplat-cli/test/framework/../recordings/arm-cli-vm-extension-tests/arm_compute_vm_Set_Chef_extension_for_the_created_vm.nock.js'
      at Function.Module._resolveFilename (module.js:336:15)
      at Function.Module._load (module.js:278:25)
      at Module.require (module.js:365:17)
      at require (module.js:384:17)
      at ARMCLITest._.extend.setupTest (/home/travis/build/Azure/azure-xplat-cli/test/framework/cli-test.js:294:20)
      at Context.<anonymous> (/home/travis/build/Azure/azure-xplat-cli/test/commands/arm/vm/arm.vm.extension-tests.js:93:13)
      at Hook.Runnable.run (/home/travis/build/Azure/azure-xplat-cli/node_modules/mocha/lib/runnable.js:204:15)
      at next (/home/travis/build/Azure/azure-xplat-cli/node_modules/mocha/lib/runner.js:263:10)
      at Immediate._onImmediate (/home/travis/build/Azure/azure-xplat-cli/node_modules/mocha/lib/runner.js:280:5)
      at processImmediate [as _immediateCallback] (timers.js:367:17)

**Comment by siddheshwar-more** [comment]
@yugangw-msft 
All tests recording done. Please merge PR.

Thank you for help!!!

**Comment by yugangw-msft** [comment]
@huangpf , please take a look

**Comment by amarzavery** [comment]
@siddheshwar-more Please take a look at the CI job. It just hung at your tests in playback mode
```
 arm compute vm Set Chef extension for the created vm:   1) arm compute vm Set Chef extension for the created vm
    arm compute vm Extension Get-Chef should list only chef extension:   2) arm compute vm Extension Get-Chef should list only chef extension
    arm compute vm Uninstall the set extension: 
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

File: lib/commands/arm/vm/vm._js [hunk 1]
```diff
@@ -532,6 +532,31 @@ exports.init = function (cli) {
         vmClient.setExtension(resourceGroup, vmName, name, publisherName, version, options, _);
     });
 
+    extension.command('set-chef [resource-group] [vm-name]')
+      .description($('Enable/disable chef extension for a VM in a resource group'))
+      .usage('[options] <resource-group> <vm-name> ')
+      .option('-g, --resource-group <resource-group>', $('the resource group name'))
+      .option('-m, --vm-name <vm-name>', $('the virtual machine name'))
```
Comment (amarzavery) [inline lib/commands/arm/vm/vm._js]: This is what is being used in other commands:
`.option('-n, --vm-name <vm-name>', $('the virtual machine name'))` Is there  a particular reason why you changed tho to `m`? Please be consistent with shorthand switches. Make sure that other switch names are also in sync.

File: lib/commands/arm/vm/vm._js [hunk 2]
```diff
@@ -532,6 +532,31 @@ exports.init = function (cli) {
...
+      .option('-o, --version <version>', $('the extension version'))
+      .option('-R, --run-list <run-list>', $('Runlist of roles/recipes to apply to VM'))
+      .option('-O, --validation-pem <validation-pem>', $('chef validation pem file path. Not required if -C or --client-pem option specified'))
+      .option('-c, --client-config <client-config>', $('chef client configuration file(i.e client.rb) path'))
+      .option('-C, --client-pem <client-pem>', $('chef client pem file path i.e required in validator less bootstrap'))
+      .option('-a, --auto-update-client', $('auto update chef client'))
+      .option('-j, --bootstrap-options <bootstrap-json-attribute>', $('Bootstrap options in JSON format. Ex: -j \'{"chef_node_name":"test_node"}\''))
+      .option('-D, --delete-chef-config', $('delete chef config files during update/uninstall extension'))
+      .option('-u, --uninstall', $('uninstall extension'))
+      .option('-t, --tags <tags>', $('Tags to set to the resource group. Can be mutliple. ' +
```
Comment (huangpf) [inline lib/commands/arm/vm/vm._js]: mutliple => multiple

File: test/commands/arm/vm/arm.vm.extension-tests.js [hunk 3]
```diff
@@ -48,6 +48,8 @@ var groupName,
   IaasDiagVersion,
   bgInfoExtension = 'BGInfo',
   bgInfoExtensionVersion = '2.1',
+  clientConfig = 'test/data/set-chef-extension-client-config.rb',
```
Comment (huangpf) [inline test/commands/arm/vm/arm.vm.extension-tests.js]: Do you need to check in these files?

File: test/commands/arm/vm/arm.vm.extension-tests.js [hunk 4]
```diff
@@ -176,6 +178,29 @@ describe('arm', function() {
         });
       });
 
+      //Set chef extension
+      it('Set Chef extension for the created vm', function(done) {
+        var cmd = util.format('vm extension set-chef %s %s --client-config %s --validation-pem %s --json', groupName, vmPrefix, clientConfig, validationPem).split(' ');
```
Comment (huangpf) [inline test/commands/arm/vm/arm.vm.extension-tests.js]: Just FYI. It seems the test coverage is only on 2 out of about 10 parameters.

File: test/recordings/arm-cli-vm-extension-tests/arm_compute_vm_Extension_Get-Chef_should_list_only_chef_extension.nock.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
+// This file has been autogenerated.
+
+var profile = require('../../../lib/util/profile');
+
+exports.getMockedProfile = function () {
+  var newProfile = new profile.Profile();
+
+  newProfile.addSubscription(new profile.Subscription({
+    id: 'e33f361b-53c2-4cc7-b829-78906708387b',
```
Comment (huangpf) [inline test/recordings/arm-cli-vm-extension-tests/arm_compute_vm_Extension_Get-Chef_should_list_only_chef_extension.nock.js]: How did you get this? :-)

**Comment by huangpf** [comment]
LGTM, if the CI tests pass.

File: test/recordings/arm-cli-vm-extension-tests/arm_compute_vm_Extension_Get-Chef_should_list_only_chef_extension.nock.js (hunk 5 above)
Comment (siddheshwar-more) [inline test/recordings/arm-cli-vm-extension-tests/arm_compute_vm_Extension_Get-Chef_should_list_only_chef_extension.nock.js]: I've replaced my subscription id with old one!

File: test/commands/arm/vm/arm.vm.extension-tests.js (hunk 3 above)
Comment (siddheshwar-more) [inline test/commands/arm/vm/arm.vm.extension-tests.js]: @huangpf These files already added at the time of chef extension `asm` implementation.

Comment (huangpf) [inline test/commands/arm/vm/arm.vm.extension-tests.js]: OK. Thanks.

File: lib/commands/arm/vm/vm._js (hunk 1 above)
Comment (siddheshwar-more) [inline lib/commands/arm/vm/vm._js]: @amarzavery I was referring to existing `azure vm extension set` commands options. [Refer](https://github.com/Azure/azure-xplat-cli/blob/dev/lib/commands/arm/vm/vm._js#L510) Where, 
```
option('-m, --vm-name <vm-name>', $('the virtual machine name'))
```
Let me know I can change both options ?

Thank you.

Comment (siddheshwar-more) [inline lib/commands/arm/vm/vm._js]: I think only for vm extension commands `-n` is used for `--extension-name`.

**Comment by amarzavery** [comment]
@siddheshwar-more Posting this gain to bring it to your notice. Please take a look at the CI job. It just hung at your tests in playback mode
```
 arm compute vm Set Chef extension for the created vm:   1) arm compute vm Set Chef extension for the created vm
    arm compute vm Extension Get-Chef should list only chef extension:   2) arm compute vm Extension Get-Chef should list only chef extension
    arm compute vm Uninstall the set extension: 
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
```

Until the CI does not pass this PR will not be merged. The codefreeze for current release is EOD today. If the PR does not get through today then these changes will go in the January 12 release of xplat cli,

**Comment by siddheshwar-more** [comment]
@amarzavery  
Yeah we are debugging an issue. I'm wondering why this hanging in tests as its works fine with actual command. Now, have feeling recording tests is more difficult than writing code :D  :) . Anyways we are looking into this.

**Comment by yugangw-msft** [comment]
@siddheshwar-more, I would double check whether the command launched by your test is the same one  you manually typed. If not, I suspect the command might stuck on prompt for your input.

**Comment by NimishaS** [comment]
@amarzavery  I verified the `azure vm extension set-chef` command, it works fine although it takes time to complete. Somehow when I am running tests in `recording` mode, it times out for the two specs which I have skipped here https://github.com/ClogenyTechnologies/azure-xplat-cli/commit/9e0682db27a2e8c71914ef10a4e277488d81c00b. We are unable to record tests.
Please suggest.

**Comment by NimishaS** [comment]
@amarzavery, we have tried running tests in `playback` mode on ubuntu 12.04 and 14.04. It ran successfully on them. However it did fail for a particular image of ubuntu 12.04.  But on that machine we couldn't record tests also.

**Comment by NimishaS** [comment]
Failed Travis that needs to be fixed: https://travis-ci.org/Azure/azure-xplat-cli/builds/99413473

**Comment by NimishaS** [comment]
@amarzavery, any suggestions on this?

**Comment by yugangw-msft** [comment]
@NimishaS , please rebase with the latest dev branch and push. This triggers a new build for you to analyze. If you still need help, let us know.

**Comment by yugangw-msft** [comment]
and add @siddheshwar-more as well

**Comment by NimishaS** [comment]
@yugangw-msft , git rebase is done. Our main issue was that Travis was failing for 2 specs during `playback` mode. Please refer this review comment https://github.com/Azure/azure-xplat-cli/pull/2222#issuecomment-168125235

We have skipped the two specs that are failing currently: https://github.com/Azure/azure-xplat-cli/pull/2222/files#diff-46be5965696efb4a3e9518831106dd93R186

**Comment by huangpf** [comment]
Why do those tests fail in playback? Is it because it has some randomization code so that the generated values are different in record mode vs. playback? Or other causes?

**Comment by yugangw-msft** [comment]
@NimishaS  for the original timeout issue, it is caused by the "lib/plugins.arm.json" was no updated with your new commands. So please run "node bin/azure --gen arm" to get it refreshed. After that please bring the tests back, which should fail now with "test/data/set-chef-extension-client-config.rb not found". This is expected because your PR doesn't include it.
At the same time if you have Visual studio, please install [node tooling](https://www.visualstudio.com/en-us/features/node-js-vs.aspx), this way you can debug your test code which should save you bunches of time.

**Comment by NimishaS** [comment]
Thanks @yugangw-msft , that was helpful. This PR passes the Travis now. Please review.

**Comment by yugangw-msft** [comment]
@NimishaS @siddheshwar-more , thanks for the contributions!

**Pull request merged**